### PR TITLE
[NIFI-13247] - Property Verification

### PR DESCRIPTION
**Pull request by rfellows**
[NIFI-13247](https://issues.apache.org/jira/browse/NIFI-13247)

Added property verification to
* Edit parameters
* Edit controller services
* Edit management controller services
* Edit reporting tasks
* Edit flow analysis rules
* Edit parameter providers

### CODE REVIEW

**Comment by mcgilman** [comment]
Will review...

File: nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/settings/ui/reporting-tasks/edit-reporting-task/edit-reporting-task.component.ts [hunk 1]
```diff
@@ -79,6 +82,10 @@ export class EditReportingTask extends CloseOnEscapeDialog {
     @Input() goToService!: (serviceId: string) => void;
     @Input() goToReferencingComponent!: (component: ControllerServiceReferencingComponent) => void;
     @Input() saving$!: Observable<boolean>;
+    @Input() propertyVerificationResults$!: Observable<ConfigVerificationResult[]>;
+    @Input() propertyVerificationStatus$: Observable<'pending' | 'loading' | 'success'> = of('pending');
+
+    @Output() verify: EventEmitter<any> = new EventEmitter();
```
[the remark this answers is not in the document]
Comment (mcgilman) [reply nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/settings/ui/reporting-tasks/edit-reporting-task/edit-reporting-task.component.ts:88]: Thoughts on changing the type being emitted to include the `ReportingTaskEntity` and the modified properties? If we did that we wouldn't need to call this method from the Effects. This question/comment applies to all new `verify` Output emitters.

File: nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/service/property-verification.service.ts [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        ) as Observable<PropertyVerificationResponse>;
+    }
+
+    deletePropertyVerificationRequest(requestId: string, uri: string) {
+        return this.httpClient.delete(
+            `${this.nifiCommon.stripProtocol(uri)}/config/verification-requests/${requestId}`
+        );
+    }
+
+    getPropertyDescriptor(id: string, propertyName: string, sensitive: boolean): Observable<PropertyDescriptorEntity> {
```
Comment (mcgilman) [inline nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/service/property-verification.service.ts:73]: This is not used.

File: nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/ui/common/property-verification/property-verification.component.html [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+                <i class="fa fa-check"></i>
+            </button>
+        </div>
+    </div>
+    @if (disabled) {
+        <div class="verification-disabled unset">Property verification is disabled</div>
+    } @else if (isVerifying) {
+        <div class="empty">Verifying properties...</div>
+    } @else if (results.length > 0) {
+        <div class="verification-results flex flex-col gap-y-2 overflow-y-auto">
```
Comment (mcgilman) [inline nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/ui/common/property-verification/property-verification.component.html:32]: IMO I think there should be a border around the verification like we do for Service references or Fetch parameters. Happy to discuss this more. Generally it looks good but my primary concern in when the results scroll like in the pic below. The border helps when the text clips.

Screenshot 2024-05-21 at 4 14 02 PM

File: nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/ui/common/property-table/property-table.component.ts [hunk 4]
```diff
@@ -114,6 +114,8 @@ export class PropertyTable implements AfterViewInit, ControlValueAccessor {
     @Input() goToService!: (serviceId: string) => void;
     @Input() supportsSensitiveDynamicProperties = false;
     @Input() propertyHistory: ComponentHistory | undefined;
+    @Input() supportsParameters: boolean = true;
+    @Input() forReferencedAttributes: boolean = false;
```
Comment (mcgilman) [inline nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/ui/common/property-table/property-table.component.ts:118]: I don't think the changes introduced here for the `forReferencedAttributes` Input are being used. Assuming that is the case, we should be able to revert the changes in the `PropertyTable` component.

File: nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/ui/common/property-verification/property-verification.component.spec.ts [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { ComponentFixture, TestBed } from '@angular/core/testing';
+import { PropertyVerification } from './property-verification.component';
+
+describe('PropertyVerificationComponent', () => {
```
Comment (mcgilman) [inline nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/ui/common/property-verification/property-verification.component.spec.ts:21]:
```suggestion
describe('PropertyVerification', () => {
```

File: nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/ui/common/property-verification/common/referenced-attributes-dialog/referenced-attributes-dialog.component.ts [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        NifiTooltipDirective,
+        MatIconButton
+    ],
+    templateUrl: './referenced-attributes-dialog.component.html',
+    styleUrl: './referenced-attributes-dialog.component.scss'
+})
+export class ReferencedAttributesDialog extends CloseOnEscapeDialog {
+    referencedAttributesForm: FormGroup;
+
+    @Input() createNew!: (existingProperties: string[]) => Observable<MapTableEntry>;
```
Comment (mcgilman) [inline nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/ui/common/property-verification/common/referenced-attributes-dialog/referenced-attributes-dialog.component.ts:63]:
```suggestion
    @Input() createNew!: (existingEntries: string[]) => Observable<MapTableEntry>;
```

**Review by mcgilman — COMMENTED** [review]
Thanks @rfellows! Works great... just noted a few things below.

File: nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/ui/common/property-verification/property-verification.component.html (hunk 3 above)
Comment (rfellows) [reply nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/ui/common/property-verification/property-verification.component.html:32]: This makes sense. I can add a border.

File: nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/settings/ui/reporting-tasks/edit-reporting-task/edit-reporting-task.component.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (rfellows) [reply nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-frontend/src/main/nifi/src/app/pages/settings/ui/reporting-tasks/edit-reporting-task/edit-reporting-task.component.ts:88]: Good idea. I'll implement it that way.

**Review by mcgilman — APPROVED** [review]
Thanks for the updates @rfellows!
